### PR TITLE
#713 send tx from mempool worker

### PR DESCRIPTION
**Pull request by rozhkovdmitrii**

### CODE REVIEW

File: proxy/common_neon/data.py [hunk 1]
```diff
@@ -1,11 +1,19 @@
...
 from dataclasses import dataclass
 from typing import Dict, Any
+from .eth_proto import Trx as NeonTx
 
 
 @dataclass
-class NeonTxPrecheckResult:
-    is_underpriced_tx_without_chainid: bool
+class NeonTxCfg:
+    is_without_chainid: bool
```
Comment (afalaleev) [inline proxy/common_neon/data.py]: The name for variable is incorrect.

If transaction without chain id, but gas price and gas limit are enough for execution, it will be executed as other Neon transactions.

We added special instruction which can process tx without chain id and with low gas limit.

File: proxy/common_neon/utils/pickable_data_server.py [hunk 2]
```diff
@@ -1,84 +1,152 @@
...
 import struct
-from typing import Any
 from logged_groups import logged_group
 
 
 class PickableDataServerUser(ABC):
 
     @abstractmethod
-    def on_data_received(self, data: Any):
+    def on_data_received(self, data: Any) -> Any:
```
Comment (afalaleev) [inline proxy/common_neon/utils/pickable_data_server.py:15]: I think we should rename method to `on_request_received()` and parameter to `request: Any`.
And split names of structures: XXXRequest and XXXAnswer

File: proxy/neon_rpc_api_model/neon_rpc_api_model.py [hunk 3]
```diff
@@ -444,15 +444,21 @@ def eth_sendRawTransaction(self, rawTrx: str) -> str:
 
         self._stat_tx_begin()
         try:
-            neon_tx_precheck_result = self.precheck(trx)
+            emulating_result: NeonEmulatingResult = call_trx_emulated(trx)
+            neon_tx_cfg = self.precheck(trx, emulating_result)
```
Comment (afalaleev) [inline proxy/neon_rpc_api_model/neon_rpc_api_model.py]: We fixed this error previously.
You can't emulate transaction before checking values in the fields of the transaction.

Revert this changes, it will be very hard to find reasons why proxy answers an user with not-clear error.

File: proxy/neon_rpc_api_model/transaction_validator.py [hunk 4]
```diff
@@ -58,15 +57,14 @@ def is_underpriced_tx_without_chainid(self) -> bool:
...
+    def precheck(self, emulating_result: NeonEmulatingResult) -> NeonTxCfg:
         try:
             self._prevalidate_tx()
-            emulating_result: NeonEmulatingResult = call_trx_emulated(self._tx)
             self._prevalidate_emulator(emulating_result)
 
-            is_underpriced_tx_without_chainid = self.is_underpriced_tx_without_chainid()
-            precheck_result = NeonTxPrecheckResult(emulating_result=emulating_result,
-                                                   is_underpriced_tx_without_chainid=is_underpriced_tx_without_chainid)
+            is_without_chainid = self.is_underpriced_tx_without_chainid()
```
Comment (afalaleev) [inline proxy/neon_rpc_api_model/transaction_validator.py]: Bad name for the variable

File: proxy/mempool/transaction_sender.py [hunk 5]
```diff
@@ -251,7 +252,7 @@ def _validate_txsize(self) -> bool:
             raise
 
     def _validate_gas_limit(self):
-        if not self._precheck_result.is_underpriced_tx_without_chainid:
+        if not self._neon_tx_exec_cfg.is_without_chainid:
```
Comment (afalaleev) [inline proxy/mempool/transaction_sender.py]: Bad name for the variable.

File: proxy/mempool/transaction_sender.py [hunk 6]
```diff
@@ -22,12 +21,14 @@
...
 from ..common_neon.utils import NeonTxResultInfo, NeonTxInfo
 from ..common_neon.errors import EthereumError
-from ..common_neon.data import NeonTxPrecheckResult, NeonEmulatingResult
+from ..common_neon.data import NeonTxCfg, NeonEmulatingResult
 from ..environment import RETRY_ON_FAIL
 from ..environment import HOLDER_MSG_SIZE
 from ..memdb.memdb import MemDB, NeonPendingTxInfo
 from ..common_neon.utils import get_holder_msg
 
+from .operator_resource_list import OperatorResourceInfo
```
Comment (afalaleev) [inline proxy/mempool/transaction_sender.py:30]: What was the reason to change the order  of imports.

File: proxy/common_neon/utils/utils.py [hunk 7]
```diff
@@ -129,6 +126,8 @@ def fill_block_info(self, block: SolanaBlockInfo):
             rec['blockNumber'] = hex(block.slot)
 
     def decode(self, neon_sign: str, tx: {}, ix_idx=-1) -> NeonTxResultInfo:
+        # TODO: move it out from here
+        from ...environment import EVM_LOADER_ID
```
Comment (afalaleev) [inline proxy/common_neon/utils/utils.py]: You've again mixed changes with refactoring.

File: proxy/common_neon/utils/pickable_data_server.py [hunk 8]
```diff
@@ -1,84 +1,152 @@
...
+                return None
+            data_len = struct.unpack("!I", len_packed)[0]
+            data = await reader.readexactly(data_len)
+            if not data:
+                return None
+            result = pickle.loads(data)
+            return result
+        except BaseException as err:
+            self.error(f"Failed to send data: {err}")
+            raise Exception("Failed to send pickable data")
```
Comment (afalaleev) [inline proxy/common_neon/utils/pickable_data_server.py:132]: I didn't find a place which uses this method

File: proxy/common_neon/utils/pickable_data_server.py (hunk 2 above)
Comment (afalaleev) [reply proxy/common_neon/utils/pickable_data_server.py:15]: It will make code more clear.

**Review by afalaleev — CHANGES_REQUESTED** [review]
You again broke logic for tx validation.
At the first step proxy should validate fields in tx, and only after proxy can emulate it

File: proxy/common_neon/utils/pickable_data_server.py (hunk 2 above)
Comment (rozhkovdmitrii) [reply proxy/common_neon/utils/pickable_data_server.py:15]: We can't mix the sense, the only responsibility of PickableDataServer is just to notify about any received data

Comment (rozhkovdmitrii) [reply proxy/common_neon/utils/pickable_data_server.py:15]: Here is no place for any business logic

File: proxy/common_neon/utils/pickable_data_server.py (hunk 8 above)
Comment (rozhkovdmitrii) [reply proxy/common_neon/utils/pickable_data_server.py:132]: It'll be used in the MemPool to send the data in async way

File: proxy/common_neon/data.py (hunk 1 above)
Comment (rozhkovdmitrii) [reply proxy/common_neon/data.py]: solved

File: proxy/common_neon/utils/utils.py (hunk 7 above)
Comment (rozhkovdmitrii) [reply proxy/common_neon/utils/utils.py]: Have made rollback

File: proxy/mempool/transaction_sender.py (hunk 6 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/transaction_sender.py:30]: At first come common imports
At second come relative imports
At last come local (current package) imports

File: proxy/neon_rpc_api_model/neon_rpc_api_model.py [hunk 9]
```diff
@@ -444,15 +444,20 @@ def eth_sendRawTransaction(self, rawTrx: str) -> str:
...
 
             self._stat_tx_success()
-            neon_tx_data = NeonTxData(tx_signed=rawTrx)
-            self._mempool_client.send_raw_transaction(neon_tx_data)
+            mempool_tx_request = MemPoolTxRequest(neon_tx=trx,
+                                                  neon_tx_exec_cfg=neon_tx_cfg,
+                                                  emulating_result=emulating_result)
+
+            if not self._mempool_client.send_raw_transaction(mempool_tx_request):
+                raise Exception("Failed to pass neon_tx into MemPool")
```
Comment (afalaleev) [inline proxy/neon_rpc_api_model/neon_rpc_api_model.py:459]: We should fix this in the next PR.
You send answer to the Ethereum client in this point, and this is strange, that you already know reason but raise Unknown Exception which will be resolved on the high level of the RPC module.

File: proxy/mempool/transaction_sender.py (hunk 6 above)
Comment (afalaleev) [reply proxy/mempool/transaction_sender.py:30]: I disagree with refactoring in the PRs which brings new logic.
Please, don't do it in the next PRs.

**Review by afalaleev — APPROVED** [review]
